File: 144.Binary_Tree_Preorder_Travesal/Binary_Tree_Preorder_Travesal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        res = []
        node = root
        while node or stack:
            if node:
                res.append(node.val)
                if node.right:
                    stack.append(node.right)
                node = node.left
            else:
                node = stack.pop()
        return res

File: 144.Binary_Tree_Preorder_Travesal/Binary_Tree_Preorder_Travesal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode *> s;
        TreeNode *node = root;
        while(node != NULL || !s.empty()){
            if(node != NULL){
                result.push_back(node);
                if(node->right != NULL)
                    s.push(node->right);
            }
            else{
                TreeNode *tmp = s.top();
                s.pop();
                node = tmp;
            }
        }        
        return result;
    }
};
